staging/src/k8s.io/apimachinery/pkg/runtime/serializer/json/json.go

```diff
@@ -213,6 +241,20 @@ func (s *Serializer) Decode(originalData []byte, gvk *schema.GroupVersionKind, i
 		return nil, actual, err
 	}
 
+	// Handle strict mode.
+	if s.strict {
+		// Delayed failure for the strict YAML to JSON conversation,
+		// so that the error output also includes GVK.
+		if errorYAMLToJSON != nil {
+			return nil, actual, fmt.Errorf("%#v: %v", actual, errorYAMLToJSON)
+		}
+		// Return an error for unknown JSON fields in strict mode.
+		strictObj := obj.DeepCopyObject()
```

Comment:
does strict decoding *require* decoding twice and copying the object? that would make it significantly more expensive to use in API server paths

Reply:
the strict YAMLtoJSON and UnmarsalStrict calls from the k8s.yaml library will always fail on error.

i was thinking about using these calls for the actual result/output too, but this might create discrepancies between json-iter and encoding/json. i think json-iter was choose for a very good reason over encoding/json in api-machinery - possibly case sensitivity and speed?

but if you think that we can just use them as the decoders this will remove the need to make duplicate objects.

Reply:
> i was thinking about using these calls for the actual result/output too, but this might create discrepancies between json-iter and encoding/json. i think json-iter was choose for a very good reason over encoding/json in api-machinery - possibly case sensitivity and speed?

can you give more details about the case-sensitivity behavior? does this library treat `{"metadata":{"Namespace":"..."}}` as an unknown field, or does it fuzzy match it to the "namespace" field? we need it to match the behavior of the thing we're actually using to decode. 

I actually expected to use json-iter's DisallowUnknownFields option for unknown fields, and only make use of the "fail on duplicate fields" option during yaml->json conversion

Reply:
right -- alternatively, is there a way we could use json-iter directly, but with some strict configuration of it? does json-iter support erroring on unknown fields? (we already catch the duplicate case with YAMLToJSONStrict in any case)

Reply:
> both ways would be computationally similar as unmarshaling has to be done twice.
> my assumption is that we want to go for 2.

I'd actually want to see the following instead of either of those options:
1. improve the YAMLToJSON convertor to return a list of duplicate key paths encountered while converting. callers can decide whether to fail, warn, or ignore if that list is not empty
2. improve jsoniter decoding to be able to expose a list of unknown field paths encountered while decoding

that would let us avoid any additional copies or unmarshals

Reply:
Checked this and it seems to be pretty straightforward actually. We'd create a new `caseSensitiveStrictJsonIterator` and ~duplicate this code using the strict option:
```
// CaseSensitiveJsonIterator returns a jsoniterator API that's configured to be
// case-sensitive when unmarshalling, and otherwise compatible with
// the encoding/json standard library.
func CaseSensitiveJsonIterator() jsoniter.API {
	config := jsoniter.Config{
		EscapeHTML:             true,
		SortMapKeys:            true,
		ValidateJsonRawMessage: true,
		CaseSensitive:          true,
	}.Froze()
	// Force jsoniter to decode number to interface{} via int64/float64, if possible.
	config.RegisterExtension(&customNumberExtension{})
	return config
}

// Private copy of jsoniter to try to shield against possible mutations
// from outside. Still does not protect from package level jsoniter.Register*() functions - someone calling them
// in some other library will mess with every usage of the jsoniter library in the whole program.
// See https://github.com/json-iterator/go/issues/265
var caseSensitiveJsonIterator = CaseSensitiveJsonIterator()
```

Then just use this iterator in `Serializer.Decode` instead of the "normal" iterator. This way we'd avoid two decodes. I'm pretty sure the YAMLToJSONStrict code supports both YAML and JSON inputs (as YAML is a superset of JSON), so we could use this code in ALL cases (not just if you specify YAML) for detecting duplicates in the original data. I think this could be the best option we have so far.

Reply:
> Checked this and it seems to be pretty straightforward actually.

json iter does suppot it, yes. that was my first attempt.
with the delayed duplicate error approch we can now use https://github.com/json-iterator/go/blob/master/config.go#L135

and create a separate `caseSensitiveStrictJsonIterator` this will avoid desrepencies between the parsers.

Reply:
@liggitt 
> -improve the YAMLToJSON convertor to accumulate a list of duplicate key paths encountered while converting. callers can decide whether to fail, warn, or ignore if that list is not empty

the yaml to json strict mode comes from the yaml.2 library:
https://github.com/kubernetes-sigs/yaml/blob/master/yaml.go#L124

here:
https://github.com/go-yaml/yaml/blob/v2.2.2/yaml.go#L88

> - improve jsoniter decoding to be able to accumulate a list of unknown field paths encountered while decoding
that would let us avoid any additional copies or unmarshals

unfortunately, these are a couple of upstream changes that may or may not be accepted and i don't want to push for them.

we also have the option to go for 1 but with better error definitions on our side.
i.e. make MyCodecFactory.UniversalStrictDecoder() exit with a meaningful error but not finish the unmarshal.